File: src/clients/python/pysharpcoredb/src/pysharpcoredb/ws_client.py
"""WebSocket client implementation for PySharpDB."""

import asyncio
import json
import logging
from typing import Any, Dict, List, Optional

import websockets
from websockets.exceptions import ConnectionClosedError, WebSocketException

from .exceptions import ConnectionError, AuthenticationError, QueryError
from .types import ConnectionInfo, ResultSet, Column, Row, ParameterValue

logger = logging.getLogger(__name__)


class WebSocketClient:
    """WebSocket client for SharpCoreDB Server streaming operations."""

    def __init__(self, host: str, port: int, tls: bool = True, timeout: float = 30.0):
        self.host = host
        self.port = port
        self.tls = tls
        self.timeout = timeout

        self._websocket: Optional[websockets.WebSocketServerProtocol] = None
        self._session_id: Optional[str] = None
        self._database: Optional[str] = None
        self._message_id = 0

        # Response queues for async operations
        self._response_queues: Dict[int, asyncio.Queue] = {}

    async def connect(self, database: str = "default", username: Optional[str] = None,
                     password: Optional[str] = None) -> ConnectionInfo:
        """Establish WebSocket connection and authenticate."""
        self._database = database

        uri = f"{'wss' if self.tls else 'ws'}://{self.host}:{self.port}/ws"

        try:
            self._websocket = await websockets.connect(
                uri,
                extra_headers=self._get_headers(username, password),
                open_timeout=self.timeout
            )

            # Start message handler
            asyncio.create_task(self._message_handler())

            # TODO: Send authentication message if needed
            # For now, assume connection is authenticated

            return ConnectionInfo(
                database_name=database,
                server_version="1.5.0"  # TODO: Get from server
            )

        except (WebSocketException, asyncio.TimeoutError) as e:
            raise ConnectionError(f"Failed to connect to WebSocket: {e}")

    async def disconnect(self) -> None:
        """Close the WebSocket connection."""
        if self._websocket:
            await self._websocket.close()
            self._websocket = None
        self._session_id = None
        self._database = None
        self._response_queues.clear()

    async def execute_query(self, sql: str, parameters: Optional[Dict[str, ParameterValue]] = None) -> ResultSet:
        """Execute a SELECT query via WebSocket."""
        if not self._websocket or not self._database:
            raise ConnectionError("Not connected")

        message_id = self._get_next_message_id()
        queue = asyncio.Queue()
        self._response_queues[message_id] = queue

        message = {
            "id": message_id,
            "type": "query",
            "sql": sql,
            "database": self._database
        }

        if parameters:
            message["parameters"] = parameters

        try:
            await self._websocket.send(json.dumps(message))

            # Wait for response
            response = await asyncio.wait_for(
                queue.get(),
                timeout=self.timeout
            )

            if response.get("error"):
                raise QueryError(response["error"], sql=sql)

            # Parse response
            data = response.get("data", {})
            columns = [
                Column(name=col["name"], type_name=col["type"], nullable=col.get("nullable", True))
                for col in data.get("columns", [])
            ]
            rows = [Row(values=row) for row in data.get("rows", [])]

            return ResultSet(
                columns=columns,
                rows=rows,
                row_count=len(rows),
                execution_time_ms=data.get("executionTimeMs", 0.0)
            )

        except asyncio.TimeoutError:
            raise QueryError("Query timeout", sql=sql)
        finally:
            self._response_queues.pop(message_id, None)

    async def execute_non_query(self, sql: str, parameters: Optional[Dict[str, ParameterValue]] = None) -> int:
        """Execute INSERT/UPDATE/DELETE via WebSocket."""
        if not self._websocket or not self._database:
            raise ConnectionError("Not connected")

        message_id = self._get_next_message_id()
        queue = asyncio.Queue()
        self._response_queues[message_id] = queue

        message = {
            "id": message_id,
            "type": "nonquery",
            "sql": sql,
            "database": self._database
        }

        if parameters:
            message["parameters"] = parameters

        try:
            await self._websocket.send(json.dumps(message))

            # Wait for response
            response = await asyncio.wait_for(
                queue.get(),
                timeout=self.timeout
            )

            if response.get("error"):
                raise QueryError(response["error"], sql=sql)

            return response.get("data", {}).get("rowsAffected", 0)

        except asyncio.TimeoutError:
            raise QueryError("Non-query timeout", sql=sql)
        finally:
            self._response_queues.pop(message_id, None)

    async def ping(self) -> float:
        """Ping the server via WebSocket."""
        if not self._websocket:
            raise ConnectionError("Not connected")

        import time
        start_time = time.time()

        message_id = self._get_next_message_id()
        queue = asyncio.Queue()
        self._response_queues[message_id] = queue

        try:
            await self._websocket.send(json.dumps({
                "id": message_id,
                "type": "ping"
            }))

            # Wait for pong response
            response = await asyncio.wait_for(
                queue.get(),
                timeout=self.timeout
            )

            end_time = time.time()
            return (end_time - start_time) * 1000

        except asyncio.TimeoutError:
            raise ConnectionError("Ping timeout")
        finally:
            self._response_queues.pop(message_id, None)

    async def _message_handler(self) -> None:
        """Handle incoming WebSocket messages."""
        try:
            while self._websocket:
                try:
                    message = await self._websocket.recv()
                    data = json.loads(message)

                    message_id = data.get("id")
                    if message_id in self._response_queues:
                        await self._response_queues[message_id].put(data)

                except ConnectionClosedError:
                    logger.info("WebSocket connection closed")
                    break
                except json.JSONDecodeError as e:
                    logger.error(f"Invalid JSON message: {e}")
                    continue

        except Exception as e:
            logger.error(f"Message handler error: {e}")

    def _get_next_message_id(self) -> int:
        """Get next message ID."""
        self._message_id += 1
        return self._message_id

    def _get_headers(self, username: Optional[str], password: Optional[str]) -> Dict[str, str]:
        """Get WebSocket headers for authentication."""
        headers = {
            "User-Agent": "PySharpDB/1.5.0"
        }

        # TODO: Add authentication headers if needed
        if username and password:
            # Could add basic auth or JWT
            pass

        return headers

    @property
    def is_connected(self) -> bool:
        """Check if connected."""
        return self._websocket is not None and not self._websocket.closed

File: src/SharpCoreDBServerScriptClients/javascript/sharpcoredb-client/src/ws-client.ts
/**
 * WebSocket client implementation for @sharpcoredb/client
 */

import WebSocket from 'ws';
import { QueryResult, QueryParameters, ConnectionInfo } from './types';
import { ConnectionError, AuthenticationError, QueryError } from './errors';

/**
 * WebSocket client for SharpCoreDB Server streaming operations
 */
export class WebSocketClient {
  private _host: string;
  private _port: number;
  private _tls: boolean;
  private _timeout: number;

  private _ws?: WebSocket;
  private _sessionId?: string;
  private _database?: string;
  private _messageId = 0;

  private _responsePromises = new Map<number, {
    resolve: (value: any) => void;
    reject: (error: any) => void;
    timeout: NodeJS.Timeout;
  }>();

  constructor(host: string, port: number, options: { tls?: boolean; timeout?: number } = {}) {
    this._host = host;
    this._port = port;
    this._tls = options.tls !== false;
    this._timeout = options.timeout || 30000;
  }

  /**
   * Establish WebSocket connection and authenticate
   */
  async connect(database: string, username?: string, password?: string): Promise<ConnectionInfo> {
    this._database = database;

    const protocol = this._tls ? 'wss' : 'ws';
    const url = `${protocol}://${this._host}:${this._port}/ws`;

    return new Promise((resolve, reject) => {
      const ws = new WebSocket(url, {
        headers: this._getAuthHeaders(username, password),
        timeout: this._timeout
      });

      const connectionTimeout = setTimeout(() => {
        ws.close();
        reject(new ConnectionError(`WebSocket connection timeout: ${this._host}:${this._port}`));
      }, this._timeout);

      ws.on('open', () => {
        clearTimeout(connectionTimeout);
        this._ws = ws;
        this._sessionId = `ws-session-${Date.now()}`;

        // Set up message handler
        ws.on('message', (data: Buffer) => {
          try {
            const message = JSON.parse(data.toString());
            this._handleMessage(message);
          } catch (error) {
            // Ignore invalid JSON
          }
        });

        ws.on('error', (error) => {
          // Connection errors will be handled by the promise
        });

        ws.on('close', () => {
          this._cleanup();
        });

        resolve({
          databaseName: database,
          sessionId: this._sessionId,
          serverVersion: '1.5.0',
          connectedAt: new Date()
        });
      });

      ws.on('error', (error) => {
        clearTimeout(connectionTimeout);
        reject(new ConnectionError(`WebSocket connection failed: ${error.message}`));
      });
    });
  }

  /**
   * Close the WebSocket connection
   */
  async disconnect(): Promise<void> {
    if (this._ws) {
      this._ws.close();
      this._ws = undefined;
    }
    this._sessionId = undefined;
    this._database = undefined;
    this._cleanup();
  }

  /**
   * Execute a SELECT query
   */
  async executeQuery(sql: string, parameters?: QueryParameters): Promise<QueryResult> {
    if (!this._ws || !this._sessionId || !this._database) {
      throw new ConnectionError('Not connected');
    }

    const messageId = this._getNextMessageId();
    const message = {
      id: messageId,
      type: 'query',
      sql,
      database: this._database,
      ...(parameters && { parameters })
    };

    const response = await this._sendMessage(message);

    if (response.error) {
      throw new QueryError(response.error, sql, parameters);
    }

    const data = response.data || {};

    return {
      columns: data.columns?.map((col: any) => ({
        name: col.name,
        type: col.type,
        nullable: col.nullable !== false
      })) || [],
      rows: data.rows?.map((row: any[], rowIndex: number) => {
        const rowObj: any = {};
        data.columns?.forEach((col: any, colIndex: number) => {
          rowObj[col.name] = row[colIndex];
          rowObj[colIndex] = row[colIndex];
        });
        return rowObj;
      }) || [],
      rowCount: data.rows?.length || 0,
      executionTimeMs: data.executionTimeMs || 0
    };
  }

  /**
   * Execute INSERT/UPDATE/DELETE
   */
  async executeNonQuery(sql: string, parameters?: QueryParameters): Promise<number> {
    if (!this._ws || !this._sessionId || !this._database) {
      throw new ConnectionError('Not connected');
    }

    const messageId = this._getNextMessageId();
    const message = {
      id: messageId,
      type: 'nonquery',
      sql,
      database: this._database,
      ...(parameters && { parameters })
    };

    const response = await this._sendMessage(message);

    if (response.error) {
      throw new QueryError(response.error, sql, parameters);
    }

    return response.data?.rowsAffected || 0;
  }

  /**
   * Ping the server
   */
  async ping(): Promise<number> {
    if (!this._ws || !this._sessionId) {
      throw new ConnectionError('Not connected');
    }

    const startTime = Date.now();
    const messageId = this._getNextMessageId();
    const message = {
      id: messageId,
      type: 'ping'
    };

    await this._sendMessage(message);
    return Date.now() - startTime;
  }

  /**
   * Send a message and wait for response
   */
  private _sendMessage(message: any): Promise<any> {
    return new Promise((resolve, reject) => {
      if (!this._ws) {
        reject(new ConnectionError('WebSocket not connected'));
        return;
      }

      const messageId = message.id;
      const timeout = setTimeout(() => {
        this._responsePromises.delete(messageId);
        reject(new QueryError('Request timeout'));
      }, this._timeout);

      this._responsePromises.set(messageId, { resolve, reject, timeout });

      try {
        this._ws.send(JSON.stringify(message));
      } catch (error) {
        clearTimeout(timeout);
        this._responsePromises.delete(messageId);
        reject(new ConnectionError(`Failed to send message: ${error}`));
      }
    });
  }

  /**
   * Handle incoming messages
   */
  private _handleMessage(message: any): void {
    const messageId = message.id;
    const promise = this._responsePromises.get(messageId);

    if (promise) {
      clearTimeout(promise.timeout);
      this._responsePromises.delete(messageId);

      if (message.error) {
        promise.reject(new Error(message.error));
      } else {
        promise.resolve(message);
      }
    }
  }

  /**
   * Clean up pending promises
   */
  private _cleanup(): void {
    for (const [messageId, promise] of this._responsePromises) {
      clearTimeout(promise.timeout);
      promise.reject(new ConnectionError('Connection closed'));
    }
    this._responsePromises.clear();
  }

  /**
   * Get next message ID
   */
  private _getNextMessageId(): number {
    this._messageId = (this._messageId + 1) % 0xFFFFFFFF;
    return this._messageId;
  }

  /**
   * Get authentication headers
   */
  private _getAuthHeaders(username?: string, password?: string): Record<string, string> {
    const headers: Record<string, string> = {
      'User-Agent': '@sharpcoredb/client/1.5.0'
    };

    // TODO: Add authentication headers if needed
    if (username && password) {
      // Could add basic auth or JWT
    }

    return headers;
  }

  /**
   * Check if connected
   */
  get isConnected(): boolean {
    return this._ws !== undefined &&
           this._ws.readyState === WebSocket.OPEN &&
           this._sessionId !== undefined;
  }
}
